### ISSUE TITLE
Incremental package update

### ISSUE THREAD

**Issue by abraverm** [nix-community/home-manager#2698]
TL;DR
When  home-manager is used as a module in NixOS then `nixos-rebuild switch` causes to update all packages managed by home-manager all together, which in return could fail because lack of storage. Incremental package update (manually or automatic) might fix the issue.

My nix store shares partitions with the OS - swap, boot and home have their own partitions. My root partition size is 50 GB, OS takes about 2-5 GB and nix store is ~25 GB. I'm using home-manager and most of the packages are managed by it. Sometimes there is not enough space (20 GB?) to rebuild (`nixos-rebuild switch`) - as the process first downloads **ALL** the cached packages and builds some. I don't know if its because the packages are managed by home-manager or its how nixos-rebuild works.

Nix allows having two packages of different versions to exist in parallel, so why does the update must be a one-shot? 
Another workarond would be to use a temporary location for downloaded packages / build location - but this is overhead for the daily user.

**Comment by SuperSandro2000** [comment]
> why does the update must be a one-shot?

Because the revision of nixpkgs is updated from an older to a newer commit and choosing any arbitrary commit in-between could be fatal and break everything.

**Comment by abraverm** [comment]
> Because the revision of nixpkgs is updated from an older to a newer commit and choosing any arbitrary commit in-between could be fatal and break everything
Isn't it a repository management issue? Have temporally two repositories - one pointing at the old commit, and the second at the new commit. Then iterate over user packages and update the nix store with garbage cleaning and removal of old generations if needed. When finished, remove the old commit repository. Nix store should be valid after each iteration, no?

> If you update more often you should get generally less updates at once.
I have auto updates configured and it partially helps, but it seems this is not enough in case of a kernel update (or maybe it is an other package, glibc?). It seems that nixos keeps two versions of each package (and dependencies) until reboot. I think this could be improved too, maybe an option to postpone none critical updates to after reboot.

**Comment by berbiche** [comment]
Sorry, this isn't something Home Manager can automatically fix without support from Nix.

Currently, Home Manager creates a derivation that contains a standard distribution folder hierarchy with your configured files, binaries, etc.
To build this derivation, Home Manager needs to build every dependency so it must download everything.

It's possible to stop the rebuild while its ongoing and to delete older packages in your store, but you won't be able to delete anything your system currently uses.

**Comment by SuperSandro2000** [comment]
> Isn't it a repository management issue?

This is a design choice in nix.

> Have temporally two repositories - one pointing at the old commit, and the second at the new commit. Then iterate over user packages and update the nix store with garbage cleaning and removal of old generations if needed. When finished, remove the old commit repository.

This would mean you would need to eval nixpkgs a ton of times which is very resource intensive. Also you cannot always update each package one by one especially when they are strict dependencies of stuff or very core like glibc or bash.

> Nix store should be valid after each iteration, no?

Nix store is valid on most if not all commits but the output might be invalid and non functional.

> It seems that nixos keeps two versions of each package (and dependencies) until reboot. I think this could be improved too, maybe an option to postpone none critical updates to after reboot.

It needs to keep the booted kernel or the machine more or less would crash instantly. This functions the same for Debian based systems. 

---


I think you want a bigger root disk or clean up other things instead.
